### PR TITLE
[fix] support varbinary in key partition table

### LINKED ISSUE #28: [Bug]: varbinary are not support in 3.x key partition table of observer 3.x
### Check Before Asking

- [X] Please check the issue list and confirm this bug is encountered for the first time.
- [X] Please try full text in English and attach precise description.

### Environment

```markdown
opensource observer 3.1
```

### Fast Reproduce Steps

run with test test_partition_varbinary in tests

### Actual Behavior

insert the row key into the wrong partition

### Expected Behavior

_No response_

### Other Information

_No response_

### PR DESCRIPTION
**Pull request by IHEII**
### What changes were proposed in this pull request?

- fix #28, support varbinary as rowkey in the key partition table, also add more test cases and change the method to deal with utf8mb4_general_ci.

### Why are the changes needed?

- support partition table

### Will break the compatibility? How if so?

- Data that is inserted by the old client may not be found by the latest client.

### Does this PR introduce any user-facing change?

- Data that is inserted by the old client may not be found by the latest client.

### How was this patch tested?

- pass all the test 

### Checklist

- [x] I've run the tests to see all new and existing tests pass.
- [x] If this Pull Request resolves an issue, I linked to the issue in the text above.
- [x] I've informed the technical writer about the documentation change if necessary.

### CODE REVIEW

File: src/rpc/util/hash/ob_hash_sort_utf8mb4.rs [hunk 1]
```diff
@@ -3278,6 +3284,61 @@ impl ObHashSortUtf8mb4 {
...
+                return OB_CS_ILSEQ;
+            }
+            *wc = ((c & 0x07) as u32) << 18 | ((c1 ^ 0x80) as u32) << 12 | ((c2 ^ 0x80) as u32) << 6 | (c3 ^ 0x80) as u32;
+            4
+        } else {
+            OB_CS_ILSEQ
+        }
+    }
+
+    fn ob_mb_wc_utf8mb4_normal(wc: &mut u32, s: &[u8], start: i32, end: i32) -> i32 {
```
Comment (WeiXinChan) [inline src/rpc/util/hash/ob_hash_sort_utf8mb4.rs:3319]: you may add some ut test cases

File: src/rpc/util/hash/ob_hash_sort_utf8mb4.rs [hunk 2]
```diff
@@ -3278,6 +3284,61 @@ impl ObHashSortUtf8mb4 {
             if !(is_continuation_byte!(c1)) {
                 return OB_CS_ILSEQ;
             }
+            *wc = ((c & 0x1F) as u32) << 6 | (c1 ^ 0x80) as u32;
+            2
+        } else if c < 0xF0 {
```
Comment (WeiXinChan) [inline src/rpc/util/hash/ob_hash_sort_utf8mb4.rs:3289]: test it !

**Review by WeiXinChan — COMMENTED** [review]
need some test cases

File: src/rpc/util/hash/ob_hash_sort_utf8mb4.rs (hunk 2 above)
Comment (IHEII) [reply src/rpc/util/hash/ob_hash_sort_utf8mb4.rs:3289]: done ouo
add UT at the end of this file

File: src/rpc/util/hash/ob_hash_sort_utf8mb4.rs (hunk 1 above)
Comment (IHEII) [reply src/rpc/util/hash/ob_hash_sort_utf8mb4.rs:3319]: done
add UT at the end of this file

**Comment by WeiXinChan** [comment]
LGTM

**Pull request merged**